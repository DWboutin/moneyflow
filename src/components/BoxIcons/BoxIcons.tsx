import React, { VoidFunctionComponent } from "react";
import classNames from "classnames";

export type BoxiconTypes = "regular" | "solid" | "logo";
export type BoxiconSizes = "xsmall" | "small" | "medium" | "large";
export type BoxiconRotations = "90" | "180" | "270";
export type BoxiconFlips = "horizontal" | "vertical";
export type BoxiconPulls = "left" | "right";
export type BoxiconAnimations =
  | "spin"
  | "tada"
  | "flashing"
  | "burst"
  | "fade-left"
  | "fade-right"
  | "fade-up"
  | "fade-down";

export interface Props {
  name: string;
  type?: BoxiconTypes;
  size?: BoxiconSizes;
  rotate?: BoxiconRotations;
  flip?: BoxiconFlips;
  pull?: BoxiconPulls;
  animation?: BoxiconAnimations;
  fixedWidth?: boolean;
  bordered?: boolean;
  circle?: boolean;
  animationOnHover?: boolean;
  className?: string;
}

const typePrefix = {
  regular: "bx",
  solid: "bxs",
  logo: "bxl",
};

const sizeSuffix = {
  xsmall: "xs",
  small: "sm",
  medium: "md",
  large: "lg",
};

const Boxicon: VoidFunctionComponent<Props> = ({
  name,
  type = "regular",
  size = "xsmall",
  rotate,
  flip,
  fixedWidth,
  pull,
  bordered,
  circle,
  animation,
  animationOnHover,
  className,
}) => {
  const classes = classNames({
    bx: true,
    [`${typePrefix[type]}-${name}`]: true,
    [`bx-${sizeSuffix[size]}`]: true,
    [`bx-rotate-${rotate}`]: rotate,
    [`bx-flip-${flip}`]: flip,
    "bx-fw": fixedWidth,
    "bx-border": bordered && !circle,
    "bx-border-circle": bordered && circle,
    [`bx-${animation}`]: animation && !animationOnHover,
    [`bx-${animation}-hover`]: animation && animationOnHover,
    [`bx-pull-${pull}`]: pull,
    className,
  });

  return <i className={classes} />;
};

export default Boxicon;
