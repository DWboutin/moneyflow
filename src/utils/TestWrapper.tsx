import React, { ReactNode } from "react";
import { Provider } from "react-redux";
import { ToastContainer } from "react-toastify";
import { ThemeProvider } from "styled-components";

import store from "../store/store";
import GlobalStyle from "../styles/globalStyle";

import { theme } from "../styles/styles";

interface Props {
  children: ReactNode;
}

const TestWrapper: React.FunctionComponent<Props> = ({ children }) => (
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <ToastContainer />
    </ThemeProvider>
  </Provider>
);

export default TestWrapper;
