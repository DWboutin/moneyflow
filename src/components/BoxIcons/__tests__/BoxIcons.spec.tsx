import {
  fireEvent,
  render,
  RenderResult,
  screen,
} from "@testing-library/react";
import TestWrapper from "../../../utils/TestWrapper";

import BoxIcons, { Props } from "../BoxIcons";

describe("<BoxIcons />", () => {
  let renderResult: RenderResult;

  const renderBoxIcons = ({
    name,
    type,
    size,
    rotate,
    flip,
    fixedWidth,
    pull,
    bordered,
    circle,
    animation,
    animationOnHover,
    className,
  }: Props) => {
    renderResult = render(
      <TestWrapper>
        <BoxIcons
          name={name}
          type={type}
          size={size}
          rotate={rotate}
          flip={flip}
          fixedWidth={fixedWidth}
          pull={pull}
          bordered={bordered}
          circle={circle}
          animation={animation}
          animationOnHover={animationOnHover}
          className={className}
        />
      </TestWrapper>
    );
  };

  describe("default rendering", () => {
    beforeEach(() => {
      renderBoxIcons({
        name: "bitcoin",
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe("with all props provided", () => {
    beforeEach(() => {
      renderBoxIcons({
        name: "bitcoin",
        type: "solid",
        size: "large",
        rotate: "90",
        flip: "vertical",
        fixedWidth: true,
        pull: "right",
        bordered: true,
        circle: true,
        animation: "tada",
        animationOnHover: true,
        className: "classname-prop",
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });
});
