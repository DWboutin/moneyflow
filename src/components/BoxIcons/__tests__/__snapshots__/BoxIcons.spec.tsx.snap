// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoxIcons /> default rendering should match snapshot 1`] = `
<div>
  <i
    class="bx bx-bitcoin bx-xs"
  />
</div>
`;

exports[`<BoxIcons /> with all props provided should match snapshot 1`] = `
<div>
  <i
    class="bx bxs-bitcoin bx-lg bx-rotate-90 bx-flip-vertical bx-fw bx-border-circle bx-tada-hover bx-pull-right className"
  />
</div>
`;
